### ISSUE TITLE
nextdns cli gets in bad state: doh resolve: context deadline exceeded

### ISSUE THREAD

**Issue by timkgh** [nextdns/nextdns#379]
What does this error mean? I see lots of them in the log and DNS resolution is very slow.

`Dec  3 21:44:20 router nextdns[12043]: Query 192.168.1.23 UDP A cl4.apple.com. (qry=31/res=12) 5001ms : doh resolve: context deadline exceeded`

Is it a problem with the endpoint at `https://do-sfo-1.edge.nextdns.io#157.245.175.48,2604:a880:2:d1::195:b001`? Should nextdns cli attempt to switch endpoints?

**Context**
 - Version: 1.9.4
 - Platform: EdgeOS v2.0.9 (EdgeRouter ER-12)

**Comment by MatMaul** [comment]
I am getting this quite often recently, it's quite annoying.

Here are some interesting logs:
```
Wed Dec 16 17:44:41 2020 daemon.notice nextdns[25708]: Query 192.168.0.99 UDP A example.com. (qry=64/res=12) 5007ms : doh resolve: dial tcp [2a07:8dc0:18:0:13:d4ff:fe1c:8095]:443: i/o timeout
Wed Dec 16 17:44:41 2020 daemon.warn nextdns[25708]: Endpoint provider failed: https://router.nextdns.io?stack=dual: Get https://216.239.32.21:443?stack=dual: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
Wed Dec 16 17:44:46 2020 daemon.warn nextdns[25708]: Endpoint failed: https://dns1.nextdns.io#45.90.28.0,2a07:a8c0::: roundtrip: context deadline exceeded
Wed Dec 16 17:44:51 2020 daemon.warn nextdns[25708]: Endpoint failed: https://dns2.nextdns.io#45.90.30.0,2a07:a8c1::: roundtrip: context deadline exceeded
```
Seems like something is wrong with the client state and not the servers/connectivity since none of the nextdns endpoints :arrow_up: can be contacted.

DuckDuckGo tells me it could be related to some wrong use of cancel.
https://www.sohamkamani.com/golang/2018-06-17-golang-using-context-cancellation/
Here I can read:
> Another thing that you have to keep in mind is that the same context instance should be passed to all functions and go-routines that you would potentially want to cancel. Wrapping an already cancellable context with WithTimeout or WithCancel will lead to multiple possibilities through which your context could be cancelled, and should be avoided.

Could it be that https://github.com/nextdns/nextdns/blob/605d06daa1dd9fce71156bcddab374e59f7de6a2/proxy/proxy.go#L97 is missing a `.Background()` ?

Beware that I don't know how cancellable works in go more than this random 30mn search&grep, so it could be entirely something else.

**Comment by MatMaul** [comment]
Also here:
https://github.com/nextdns/nextdns/blob/460b02b5b1a7ebe18dc2fb5a226f443371784e18/resolver/endpoint/manager.go#L132
All other occurrences of `WithCancel/Timeout` seems to be called with a `.Background()` context.

**Comment by rs** [comment]
A context is a tree, background is the root. When you inherit from an existing context, you don't use background, otherwise you break the chain.

**Comment by rs** [comment]
The deadline exceeded means there were a timeout. Timeouts can happen for many reasons.

**Comment by MatMaul** [comment]
Ok so completely unrelated thx :) 

What do you think about the fact that even router.nextdns.io is unavailable ? I was first blaming unavailable servers or connectivity on my side but disabling NextDNS always fixes the problem, so at least connectivity is ruled out and 3 of your servers down including the main ones seems unlikely I would said.

Going to restart it next time  to see if it fixes too, and increase the logs level.

Let me know if I can do more investigation somehow.

**Comment by MatMaul** [comment]
Is there a way to increase logging for debugging purpose ? `nextdns config list` doesn't give me anything related to log level.

**Comment by rs** [comment]
You can enable `log-queries`, but that's about it.

**Comment by rs** [comment]
If the daemon is stuck in some ways, it might be interesting to get a stacktrace with `nextdns trace`.

**Comment by timkgh** [comment]
@rs I also see lines like the one below:

`Dec 17 07:51:51 router nextdns[12043]: Connected 157.245.175.48:443 (con=0ms tls=0ms, )`

Is this just a logging problem or something goes wrong when it tries to establish a connection? Normally I see something like `(con=9ms tls=36ms, TLS13)`

**Comment by rs** [comment]
That's odd, could be an unhandled connection error.

**Comment by maghuro** [comment]
Same is happening right now!

**Comment by timkgh** [comment]
Indeed, it does get into this state where it takes a long time to recover.

I learned long time ago that you cannot rely on a single DNS provider, it's good to have something else to fallback on, e.g. Google, Cloudflare or even ISP DNS.

**Comment by zitstif** [comment]
I'm having issues with the nextdns windows client. It's in a faulty state and in Windows eventlog under applications I'm seeing a bunch of error messages like this:

Endpoint failed: dns2.nextdns.io: x509: certificate signed by unknown authority

Endpoint failed: ipv4-vultr-chi-1.edge.nextdns.io: x509: certificate signed by unknown authority

Endpoint failed: dns2.nextdns.io: x509: certificate signed by unknown authority

Endpoint failed: ipv4-vultr-chi-1.edge.nextdns.io: x509: certificate signed by unknown authority

Endpoint failed: d1xovudkxbl47e.cloudfront.net: context deadline exceeded

Internet doesn't work and if you do an nslookup, it tries to use the "192.0.2.42" address but queries times out. 

Public servers are pingable (i.e. ping 1.1.1.2) but to fix  internet connectivity (dns resolution) I have to remove the Windows NextDNS client.

System is Windows 10 on OS Build 19042.985. For AV, I'm using the built in windows defender. 

I've tried connecting the system to two different wireless networks with 2 different providers, I get the same results. 

Any ideas? I really like the NextDNS client.

**Comment by rs** [comment]
@zitstif can you please try the beta channel and see if you can reproduce?

**Comment by zitstif** [comment]
I've enabled the beta channel in the app and the internet still isn't working. Is there a beta version of nextdns that I can download and try?

**Comment by zitstif** [comment]
Nvm, I found it: https://nextdns.io/download/windows/beta  

I will give it a shot and report back.

**Comment by zitstif** [comment]
NextDNS client 2.0.3 originally produced the same results but is magically working now. I still get timeouts on doing nslookup using 192.0.2.42. Here's some new errors I'm getting in windows eventlog:

Endpoint failed: ipv4-vultr-chi-1.edge.nextdns.io: x509: certificate signed by unknown authority

resolve: 7f74 dial tcp 45.90.30.0:443: connectex: A socket operation was attempted to an unreachable host.

resolve: 220d read tcp 192.168.108.8:58103->45.90.30.0:443: wsarecv: An established connection was aborted by the software in your host machine.



Is it possible that nextdns is having any hosting issues/outages?

**Comment by rs** [comment]
Your problem is most likely related to this change: https://twitter.com/NextDNS/status/1399363914043990020

For some reason, our client on your system is not able to find the ECC CA root of letsencrypt.

Can you please confirm that `Endpoint failed: ipv4-vultr-chi-1.edge.nextdns.io: x509: certificate signed by unknown authority` was thrown out by the beta version of the client? If you reboot windows, do you get any new certificate error with the beta version (2.0.3) of the client?

**Comment by zitstif** [comment]
Yes on the workstation with the Beta version I am no longer seeing this error. When will 2.0.3 come out of beta?

**Comment by rs** [comment]
Today apparently :)

**Comment by modxmv** [comment]
I am having similar issues with "doh resolve: context deadline exceeded" running this on an Ubiquiti USG.

**Comment by rs** [comment]
@modxmv what do you get for `nextdns version`?

**Comment by it-can** [comment]
Also getting this error on my edgerouter 3 lite

```
Jun  2 08:58:43 Kantoor nextdns[8779]: Query 192.168.2.16 UDP AAAA apidata.googleusercontent.com. (qry=47/res=12) cached HTTP/2.0: doh resolve: context deadline exceeded
Jun  2 08:58:43 Kantoor nextdns[8779]: Query 192.168.2.16 UDP 65 www.googleapis.com. (qry=36/res=12) cached HTTP/2.0: doh resolve: context deadline exceeded
Jun  2 08:58:43 Kantoor nextdns[8779]: Query 192.168.2.16 UDP A apidata.googleusercontent.com. (qry=47/res=12) cached HTTP/2.0: doh resolve: context deadline exceeded
Jun  2 08:58:43 Kantoor nextdns[8779]: Query 192.168.2.16 UDP AAAA www.googleapis.com. (qry=36/res=12) cached HTTP/2.0: doh resolve: context deadline exceeded
Jun  2 08:58:43 Kantoor nextdns[8779]: Query 192.168.2.16 UDP 65 gdmf.apple.com. (qry=32/res=12) cached HTTP/2.0: doh resolve: context deadline exceeded
Jun  2 08:58:43 Kantoor nextdns[8779]: Query 192.168.2.16 UDP A www.googleapis.com. (qry=36/res=12) cached HTTP/2.0: doh resolve: context deadline exceeded
Jun  2 08:58:43 Kantoor nextdns[8779]: Query 192.168.2.16 UDP 65 apps.eastwest.se. (qry=34/res=12) 38990ms : doh resolve: context deadline exceeded
Jun  2 08:58:43 Kantoor nextdns[8779]: Query 192.168.2.16 UDP AAAA apps.eastwest.se. (qry=34/res=12) 39005ms : doh resolve: context deadline exceeded
Jun  2 08:58:44 Kantoor nextdns[8779]: Connected 45.90.28.0:443 (con=7ms tls=839ms, TCP, TLS13)
Jun  2 08:58:44 Kantoor nextdns[8779]: Switching endpoint: https://dns.nextdns.io.#45.90.28.0,2a07:a8c0::
Jun  2 08:59:50 Kantoor nextdns[8779]: Connected 45.90.28.0:443 (con=7ms tls=0ms, TCP, )
Jun  2 09:00:41 Kantoor nextdns[8779]: Connected 45.90.28.0:443 (con=32ms tls=0ms, TCP, )
Jun  2 09:01:48 Kantoor nextdns[8779]: Connected 45.90.28.0:443 (con=46ms tls=0ms, TCP, )
Jun  2 09:02:25 Kantoor nextdns[8779]: Connected 45.90.28.0:443 (con=46ms tls=0ms, TCP, )
Jun  2 09:02:58 Kantoor nextdns[8779]: Connected 45.90.28.0:443 (con=7ms tls=0ms, TCP, )
ubnt@Kantoor:~$ ls
ubnt@Kantoor:~$ nextdns version
nextdns version 1.32.2
```

**Comment by timkgh** [comment]
Something has been very wrong with NextDNS lately. Used to get about 36ms for the TLS connection, now 10x slower:
```
Jun  1 22:34:53 router nextdns[6513]: Connected 45.90.28.0:443 (con=16ms tls=398ms, TCP, TLS13)
Jun  1 22:40:26 router nextdns[6513]: Connected 45.90.28.0:443 (con=16ms tls=378ms, TCP, TLS13)
Jun  1 22:41:44 router nextdns[6513]: Connected 45.90.28.0:443 (con=14ms tls=328ms, TCP, TLS13)
Jun  1 22:42:59 router nextdns[6513]: Connected 45.90.28.0:443 (con=15ms tls=0ms, TCP, )
Jun  1 22:47:01 router nextdns[6513]: Connected 45.90.28.0:443 (con=16ms tls=334ms, TCP, TLS13)
Jun  1 22:50:39 router nextdns[6513]: Connected 45.90.28.0:443 (con=13ms tls=0ms, TCP, )
Jun  1 22:55:22 router nextdns[6513]: Connected 45.90.28.0:443 (con=14ms tls=0ms, TCP, )
Jun  1 22:59:57 router nextdns[6513]: Connected 45.90.28.0:443 (con=13ms tls=0ms, TCP, )
Jun  1 23:08:14 router nextdns[6513]: Connected 45.90.28.0:443 (con=14ms tls=331ms, TCP, TLS13)
Jun  1 23:19:51 router nextdns[6513]: Connected 45.90.28.0:443 (con=16ms tls=333ms, TCP, TLS13)
```

**Comment by rs** [comment]
@timkgh what is the hardware you are running it on?

**Comment by timkgh** [comment]
EdgeRouter ER-12 but the 36ms number I quoted is from the same hardware, been monitoring it for the past few months.

**Comment by timkgh** [comment]
But even on x86 Linux this seems to have become a problem, it used to be really fast, <10ms
```
nextdns[686]: Connected 45.90.28.0:443 (con=0ms tls=117ms, TCP, TLS13)
nextdns[686]: Connected 104.238.181.28:443 (con=1ms tls=69ms, TCP, TLS13)
```

**Comment by zitstif** [comment]
I know this might be a little off topic.. but the 2.0.3 file is getting flagged by some AV solutions as malicious:

https://www.virustotal.com/gui/file/b0b7a1408329bc3b64e56f05e71be05a4042b3e71ae271cb137d3eef0af3f7e2/detection

https://ibb.co/7bSGs08

I'm really hoping this a false positive and that the installer wasn't poisoned lol

**Comment by joshstclair** [comment]
Why did this issue close without resolution? I just setup NextDNS cli on my raspberry pi for our home wifi and last night, it took our internet down. I woke up to a flood of logs like the ones mentioned above. Restarting NextDNS and the rpi did not fix the issue. I eventually had to factory reset our router to get everything back into a usable state.

nextdns version 1.33.11

**Comment by benlongmire** [comment]
> Why did this issue close without resolution? I just setup NextDNS cli on my raspberry pi for our home wifi and last night, it took our internet down. I woke up to a flood of logs like the ones mentioned above. Restarting NextDNS and the rpi did not fix the issue. I eventually had to factory reset our router to get everything back into a usable state.
> 
> nextdns version 1.33.11

Try something along the lines of this https://github.com/nextdns/nextdns/issues/500#issuecomment-853140837

**Comment by joshstclair** [comment]
@benlongmire, I appreciate the follow up. I read through that thread and it still looks like others were still experiencing the issue after that workaround? 
After my incident, I decided to just use the NextDNS application on our individual devices, rather than route our whole network through my rpi. I was admittedly a little scarred from that experience and didn't like how I created a single point of failure.
